Perfmon System version 2

pwinston opened the pull request:
# Description
Updates to the **perfmon system** which were made while investigating and solving #1320:
1. Use monkey patching from a config file instead of decorators.
2. Config file has "sets" of callables you can toggle on and off.
3. Config file has new option to turn Qt event tracing on or off.
4. Added `kwargs` to `perf_timer` to display per-event key/values in Chrome tracing.
5. Added `add_instant_event()`per API to add vertical lines in Chrome tracing.
6. New `napari.utils.patcher.py` does the monkey patching, it's generic not perf specific.

# Type of change
- [x] Update existing feature (non-breaking change which adds functionality)

# Example 1

<img width="1184" alt="87239376-7caaa100-c3dc-11ea-9846-f0258f82b6ce" src="https://user-images.githubusercontent.com/4163446/88871922-0fbe4600-d1e7-11ea-96d9-d8e116d17688.png">

# Example 2

<img width="1069" alt="napari-slow-8k" src="https://user-images.githubusercontent.com/4163446/88872065-6592ee00-d1e7-11ea-9077-e1b230a3af2e.png">

# Notes

### New NAPARI_TRACE_FILE env var

Set this to a file path and Napari will trace to file on startup. I basically always use this and never start a trace from the debug menu anymore, although we should keep that option because it's more discoverable and useful for one-time tracing. This is how I tend to use perfmon today:

1. Set `NAPARI_TRACE_FILE` to some local `latest.json` file
2. Run Napari, do what I want, exit with Command-Q (hitting red circle to exit does not work?)
3. Drop `latest.json` into `chrome://tracing`.

So every time I run napari I get a trace file, and if I care what's in it I just throw it into Chrome.

### Checking NAPARI_PERFMON

Before we took care to check the `NAPARI_PERFMON` env var without importing perf in a few place. Now with a config file we have to check the settings in that file to know whether to trace Qt events, so it's more natural to import perf. We check whether to trace Qt events like this:

```
if perf_config and perf_config.trace_qt_events:
    # setup tracing
```

### Perfmon Config File

If `NAPARI_PERFMON=1` we turn on Qt Event tracing like before, and we do nothing else. If `NAPARI_PERFMON` is a file path however, we load that as our perfmon config file. The format of the config file is:

```
{
    "trace_qt_events": true,
    "trace_callables": [
        "chunk_loader",
        "vispy_rendering"
    ],
    "callable_lists": {
        "chunk_loader": [
            "napari.components.dims.Dims.set_point",
            "napari.components.viewer_model.ViewerModel._update_layers",
            . . . 
        ],
        "vispy_rendering": [
            "vispy.app.canvas.Canvas.set_current",
            "vispy.gloo.program.Program.draw",
            "vispy.gloo.context.GLContext.flush_commands",            
            . . . 
        ]
    }
}
```

The `PerfmonConfig` class monkey patches our `perf_timer` context object into each function or method above. This means we don't have to modify the code at all, not even to add a decorator. And it also means we can have different trace sets, even overlapping ones, that we can toggle on or off as needed. You want enough tracing to see what you want, but not so much that the overhead is annoying.

### Add kwargs to perf_timer

Now `PerfTimer` has a `kwargs` field that lets you set arbitrary key=value pairs that will be displayed into the `chrome://tracing` GUI when you click on an event. With monkey patching you can't use these, but with a manual `perf_timer` you can.

### New add_instant_event()

New `add_instant_event()` is shown in `chrome://tracing` as a vertical line. This is useful because very short events are invisible unless you zoom way in. But instant events are always visible as lines. For instance when debugging I created instant events for `MousePressed` and `MouseReleased`. There are 3-4 other graphical things in the [chrome tracing spec](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview) that we aren't using yet, need to look into those.

### "Tracing" vs. "Timing"

I've been switching to the term "tracing" based on my experience using the perf stuff. The Chrome feature is called `chrome://tracing` and tracing really is the right term. There is timing information, but just visualizing how the code executes is a big part of things, beyond the timing information. I might rename the whole module from "perf" to "tracing" at some point.

### Call Stack

The `perf_timer` wrappers clutter up the stack. If you want to debug might want to turn off perfmon.

### Patcher

The new `napari.utils.patcher.py` is actually totally generic for patching functions/methods. The perf code uses `patcher` to patch in `perf_timer` but you could use it to patch anything. Not sure if there is some library (event `wrapt` itself?) that can patch lists of callables like we are doing, but it was not hard to implement. It's only 150 lines with comments.

# How has this been tested?
- [x] Manual testing while investigation of #1320 

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

pwinston committed:
Start new branch

pwinston committed:
First cut of perfmon2 from chunk-loader branch

pwinston committed:
Add comment

pwinston committed:
Patcher from chunk-loader branch

pwinston committed:
Refine comment

pwinston committed:
Typos in comment

pwinston committed:
Better comment wording

pwinston committed:
Update patcher to not re-import over and over and related cleanup.

pwinston committed:
Better comments

pwinston committed:
Fix env var == "0" case

pwinston committed:
Merge branch 'master' into feature/perfmon2

sofroniewn reviewed (approved):
Wow, this is amazing! Just gave it a spin and everything works great. I love the config file, so great to be able to specify the callables that I want to trace, I can imagine this being incredibly useful when developing and debuging. Right now I need to use benchmarks and airspeed velocity for all profiling, which can be nice for seeing everything, but I see this as being much faster and easier to use when figuring things out the first time.

Curious how @tlambert03 thinks about this, but I guess after we merge #1183 we'll probably want to combine that config file into that approach, but that can come later. Definitely nothing that needs to hold this PR up.

> Now PerfTimer has a kwargs field that lets you set arbitrary key=value pairs that will be displayed into the GUI when you click on an event. With monkey patching you can't use these, but with a manual perf_timer you can.

When you say GUI here you mean the chrome tracing GUI right? not napari? this is how you injected your additional metadata that you wanted to view in the tracing tool right? just want to make sure I understand.

> but you could use it to patch anything

Just curious what else could you imagine we might want to patch in like this?

### final comment

I also read through the whole code and everything seemed very simple and clear - I won't claim to understand every line after this read, but it all looks very good to me and well documented. I think this is good to merge, but will let @jni and @tlambert03 review before then

pwinston committed:
Ignore duplicate targets in the config file

tlambert03 reviewed (approved):
I'm with @sofroniewn... this looks amazing & useful and the code is very clean and readable.  I also can't claim to fully follow everything that's happening here without spending more time than I currently have... but since this is still an opt-in feature, I have no issues merging this to master.

I guess one request for a follow-up PR would be a bit of a "beginners guide" to using it.  I'm new to this sort of performance tracing, so it's not totally clear to me whether the only output here is the "`latest.json`" file which should be evaluated in `chrome://tracing`, or whether there are other ways that you're using this.  I wouldn't want you to regurgitate some chrome tracing tutorial that is already available elsewhere, but some docs that point to useful resources, with a quick example workflow (e.g. create something that doesn't perform that well, monitor it, fix it, and show how you know it got fixed) would be great.

jni commented on napari/utils/patcher.py:
```diff
@@ -0,0 +1,212 @@
+        # This is probably not needed because importing a module the second
+        # time seems instant. But it's easy to cache these and possibly
+        # it's better in some way.
+        self.modules: Dict[str, Module] = {}
```
This seems like a reimplementation of [`sys.modules`](https://docs.python.org/3/library/sys.html#sys.modules)?

pwinston replied:
Now relies on sys.modules in instead of this.

jni commented on napari/utils/patcher.py:
```diff
@@ -0,0 +1,212 @@
+perf_timers, but this can be used for any type of patching.
+
+See patch_callables() below as the main entrypoint.
+"""
```
Overall, this module seems like a lot of OO overkill (OOverkill? I like it!). It defines a Module object, but Python modules are already objects, which have the required attributes (`module.__path__`, `module.__name__`, etc.). The method `patch_attribute` could easily (I think!) be implemented as a function that takes a Python module as input. It defines an Importer object, ostensibly for caching but importlib already does importing and modules are already cached in sys.modules. Finally, the Patcher class seems to be there to keep track of the Importer, but again, that's not really necessary — if using built-ins, this could just be defined as functions.

As others have pointed out, this is all internal to napari (might be good to note that by renaming this file to `_patcher.py`), so a refactor can come later, but I think it should come for sure.

pwinston replied:
It was more as functions at first but it grew in this direction. I didn't think of using `sys.modules` that would simplify things.  And then I didn't realize `module.__name__` is the "path" in my terms, it's `napari.utils.perf` or whatever.

Any class can be mechanically turned into a module where the attributes become module globals and the methods become functions. I don't think that's automatically a good idea just because it's possible. But using `sys.modules` and `module` here does make sense. And if after making those changes there just was no point to some of the classes it might make sense to get rid of them. I'll try a version 2 of it.

pwinston replied:
New version is all functions. Using sys.modules and module.__name__ and then unwinding some things, no more needs for the classes.

jni commented on napari/utils/perf/__init__.py:
```diff
@@ -12,7 +12,15 @@
 from ._utils import perf_timer, perf_func
+
+
+# If not using perfmon timers will be 100% disable with hopefully zero
```
```suggestion
# If not using perfmon timers will be 100% disabled with hopefully zero
```

jni commented on napari/utils/perf/_compat.py:
```diff
@@ -1,11 +1,11 @@
 
-
-if PYTHON_3_7:
+if PYTHON_3_7_OR_NEWER:
```
I feel like this doesn't need to be defined if it only gets used in this one spot? `sys.version_info` is pretty standard.

pwinston replied:
Yeah I will just use `sys.version_info` directly. It used to be used in multiple places but no more.

pwinston replied:
Used it directly now

jni commented on napari/utils/perf/_config.py:
```diff
@@ -1,11 +1,150 @@
+        path = Path(config_path)
+        if not path.exists():
+            raise FileNotFoundError(
+                errno.ENOENT, f"Config file {PERFMON_ENV_VAR} not found", path,
```
I feel like this should be the filename, ie `{os.getenv(PERFMON_ENV_VAR)}`?

pwinston replied:
That's on purpose, the `path` variable is the actual path. But it was kind of silly I deleted the existence check and this raise entirely. Now it fails like:
```
FileNotFoundError: [Errno 2] No such file or directory: '/Users/pbw/.napari-does-not-exist'
```
Which is probably fine. I was just trying to be nice and remind you what env var was involved, but probably people can figure that out.

jni commented on napari/utils/perf/_config.py:
```diff
@@ -1,11 +1,150 @@
+        self._patch_callables()
+        self.patched = True
+
+    def _get_callables(self, callable_list) -> List[str]:
```
What is 'callable_list'? It should be type annotated, or annotated in the docstring, or both.

jni replied:
@pwinston did you miss this comment?

pwinston replied:
I think I saw it but didn't come back to it. I changed the var name to `list_name` and added the type `str`.

jni commented on napari/utils/perf/_event.py:
```diff
@@ -1,32 +1,49 @@
+    category :str
+        Comma separated categories such has "render,update".
+    **kwargs : dict
+        Additional keyword arguments for the "args" field of the event.
```
Can you add an Attributes section with descriptions of the attributes? e.g. I was surprised to read `self.type = "X"` :joy: but it might make sense with a description.

pwinston replied:
I added attributes just for `self.type` and `self.args` and a comment for `self.type`. The "X" is just part of the `chrome://tracing` format

jni commented on napari/utils/perf/_event.py:
```diff
@@ -1,32 +1,49 @@
+        self.category = category
+        self.args = kwargs
+        self.pid = pid
+        self.tid = threading.get_ident()
```
`pid` is pretty common, but I've never heard of tid... I wonder if these could/should be renamed to `process_id` and `thread_id`?

pwinston replied:
Changed to `process_id` and `thread_id`

jni commented on napari/utils/perf/_event.py:
```diff
@@ -47,3 +64,9 @@ def duration_us(self):
+class InstantEvent(PerfEvent):
+    def __init__(self, name: str, time_ns: int, **kwargs):
+        super().__init__(name, time_ns, time_ns, **kwargs)
+        self.type = "I"  # instant event
```
What's the difference between these two event types, and if there is a class for each, why not use `type(event)`?

pwinston replied:
I added some comments about the different event types, but if you Google for "Trace Event Format" there's a google doc with the full spec, it's about 19 pages.

I'm not sure what you are proposing exactly but in general with OO it's considered much better to ask objects for information like such as `write event.type`, where the object "knows" the right answer, than to externally interrogate their type like `write "I" if type(event) == InstantEvent" else "X"` which is considered more fragile.

That said confining the details of their format to one place in the code would be nice, so maybe it is worth using `type(event)` or an enum to have a translation spot from our code to their format. I think with just "X" and "I" it's not bad as-is, but maybe if we add more even types we should consider that.

jni commented on napari/utils/perf/_timers.py:
```diff
@@ -52,24 +56,35 @@ def __init__(self):
-            self.timers[name].add(duration_ms)
-        else:
-            self.timers[name] = Stat(duration_ms)
+        if event.type == "X":
```
ie here you would write `if type(event) == PerfEvent`

pwinston replied:
Addressed above.

jni commented on napari/utils/perf/_trace_file.py:
```diff
@@ -71,22 +84,33 @@ def write_event(self, event: PerfEvent) -> None:
-            "cat": event.category,
-            "ph": "X",
+            "cat": category,
+            "ph": event.type,
```
Ah, this answers my questions above. What is `ph` short for and is there a better description we could use for our attribute?

pwinston replied:
`ph` is `phase type` but in the spec they define it like this `ph: The event type.` so that's why we call it `event.type`, we could call it `event.phase` I guess "type" is kind of overloaded in Python. The whole definition from their spec:

>  ph: The event type. This is a single character which changes depending on the type of event being output. The valid values are listed in the table below. We will discuss each phase type below.

pwinston replied:
I changed it to `event.phase` I think that's much clearer. Both `event` and `type` are very overloaded so `event.type` was confusing. They do call it "event type" in places, but phase is the official name.

jni commented on napari/utils/perf/_trace_file.py:
```diff
@@ -71,22 +84,33 @@ def write_event(self, event: PerfEvent) -> None:
+            #     "g" - global
+            #     "p" - process
+            #     "t" - thread
+            data["s"] = "p"
```
It's unclear to me what scope means in this context or why it would always be p here?

pwinston replied:
The "scope" here is strictly a chrome://tracing term. It's saying what type of "instant event" is this? Is it an event that applies only to one thread, to one process, or to the whole system. It basically just comes down to when `chrome://tracing` draws the "instant event" where does it draw on the line.

I just hardcoded it to process based on how I was using it. Probably we should expose all 3 options since it's not hard. I just didn't even experiment with them to see what they looked like, so I didn't want to expose them without at least trying them.

pwinston replied:
I added a better comment.

jni commented on napari/utils/perf/_utils.py:
```diff
@@ -1,30 +1,38 @@
-"""Utilities to support performance monitoring:
+"""Performance monitoring utilities.
```
meta-utils! :joy:

jni commented on napari/utils/perf/_utils.py:
```diff
@@ -81,29 +96,26 @@ def draw(self):
         def decorator(func):
             @functools.wraps(func)
-            def time_function(*args, **kwargs):
+            def wrapper(*args, **kwargs):
```
I like the previous name better, as it's more descriptive... But maybe `timed_function`?

jni replied:
Also, the multiple nesting can be avoided with `toolz.curry`, which we already depend on and use in the `thread_worker` decorator:

```python
@tz.curry
def perf_func_named(timer_name : str, func : Callable) -> Callable:
    @functools.wraps(func)
    def timed_func(*args, **kwargs):
        with perf_timer(timer_name, "decorator"):
            return func(*args, **kwargs)
    return timed_func
```

By the way, what is the second argument, 'decorator', to perf_timer?

jni replied:
... And you could even combine things with the above non-named decorator:

```python
@tz.curry
def perf_func(func : Callable, timer_name : str = None) -> Callable:
    if timer_name is None:
        timer_name = f"{func.__qualname__}"

    @functools.wraps(func)
    def timed_func(*args, **kwargs):
        with perf_timer(timer_name, "decorator"):
            return func(*args, **kwargs)
    return timed_func
```

Then you would use

```python
func = perf_func(func, timer_name=my_name)
```

(using keyword arguments for `timer_name`) when wanting to overwrite the name, but just `func = perf_func(func)` when not.

I'm not sure what `full_name` is above but the same principle applies.

pwinston replied:
The "decorator" string is the second argument of `perf_timer` which is "category", which is like a comma separated list of strings that supposed let you toggle on and off events by "category". Like "rendering" vs. "IO" maybe? I've never actually seen it in the UI. I should probably just have no category until/unless we determine it's useful for something.

I've never used `toolz.curry` so that would take some investigation, but sounds potentially useful. Same with collapsing it down to less decorators.

pwinston replied:
I just deleted `perf_func` and `perf_func_named`. Those came long before we had the patcher. With the patcher there's really no reason to have these. The patcher patches in `perf_timers` it does not use these decorators. And patching using the patcher is really better in every way that using these decorators except you lose the "name" part. But we could actually add that to the patcher if we wanted.

That is the patcher config file could support something like:

    ("module1.Class2.method2", "redraw everything")

meaning patch `method2` to give the timer the name "redraw everything".

So if we ever need names like that I'd add them with the patcher.

jni commented on napari/utils/perf/_utils.py:
```diff
@@ -81,29 +96,26 @@ def draw(self):
+    # contextlib.nullcontext does not work with kwargs?
+    @contextlib.contextmanager
+    def perf_timer(name: str, category: Optional[str] = None, **kwargs):
+        yield
```
Does this context ever get used? I thought if perfmon was off there were zero runtime modifications?

pwinston replied:
I used to assign it to `contextlib.contextmanager` when disabled but I had to switch to this then I added the `kwargs` argument. I would hope Python could optimize away a generator that does absolutely nothing, but I'm not super optimistic about that, I'll do some timings.

With the patcher now I use the patcher for 99% of things. So it simply doesn't patch in the `perf_timer` at all when disabled. But yeah if you leave a raw `perf_timer` in the code at all times, this null-generator will run. So I'll see if I can measure it.

jni replied:
> But yeah if you leave a raw perf_timer in the code at all times

But we don't do this, right? I'd prefer to not enable this kind of pattern...

pwinston replied:
There's no `perf_timers` checked in except one in the class `QApplicationWithTracing` but that is separately disabled. So there's no "active" `perf_timers` checked in as of this PR. The developer is expected to add them using the patcher. 

However there's two reasons you might want to temporarily add a `perf_timer` by hand while doing an intensive investigation of something:

1. You want to time a line of code or a block of code within a bigger function. And you don't (yet) want to refactor that code to extract that block into its own function that you can patch.

2. You want to add arguments to the timer. This is really for "tracing" not such much "timing". For example you could put the `shape` of an array as an argument to a `perf_timer` or the filename of what chunk you are loading, then in the chrome GUI you can see those values and that might help explain the timeline.

So I'd consider `perf_timers` like anything you'd add temporarily while debugging, like noisy print statement. Something you should remove before you commit or at least **before** you merge. This should be covered in the documentation I haven't written. Basically explain patching is "free" but manually added `perf_timers` are not completely free. I timed a "do nothing" generator as taking like a microsecond.

If there was a problem with people commiting and merging `perf_timers` we could write a CI step that checked for them, but I wouldn't bother unless it becomes a problem.

jni commented on napari/viewer.py:
```diff
@@ -161,3 +161,15 @@ def close(self):
+    # `gui_qt` we already have the special one, and no conversion is done
+    # here. However when running inside IPython or Jupyter this is where we
+    # switch out the QApplication.
+    return convert_app_for_tracing(app)
```
Why did you pull this out into a pass-through function...? The deleted code above seems clearer than this, which adds another layer of indirection...

pwinston replied:
I agree the one line `_convert_app_for_tracing` is not needed. I think I had it doing more at first and it whittled down to just the one line. I'll get rid of it.

pwinston replied:
I removed it.

jni reviewed (commented):
@pwinston I've made quite a few comments. They are all about code style and not functionality (although I did ask a question about the "empty" context manager. Also, is there any testing of all this? Given the changes in this PR I would expect some tests to change/be added, but there's no tests here.

Happy to postpone the suggested changes to later PRs, although happier still if they could happen here, since my experience is that "later PRs" often never come...!

Also, there may well be complexities I haven't appreciated in some of my comments — if so, please feel free to point them out!

Thanks!!!

pwinston committed:
Update napari/utils/perf/__init__.py

Co-authored-by: Juan Nunez-Iglesias <juan.nunez-iglesias@monash.edu>

pwinston committed:
Simplify, rely on sys.modules

Use built-in modules class
instead of our own.

pwinston committed:
Move patcher to _perf directory

It's not perf specific, but perf is the only one using it so keep it hidden
for now, we can always move it later.

pwinston committed:
Merge branch 'feature/perfmon2' of github.com:pwinston/napari into feature/perfmon2

pwinston committed:
remove raise FileNotFoundError

Was not really needed, the open will raise the same exception.

pwinston committed:
Better comment

pwinston committed:
Remove the single-line _convert_app_for_tracing() function

pwinston committed:
Use sys.version.info directly

jni commented on napari/utils/perf/_patcher.py:
```diff
@@ -0,0 +1,155 @@
+
+
+def _patch_attribute(module, attribute_str, patch_func):
+    """Patch the module's callable pointed to by the attribute string.
```
This needs a full docstring.

pwinston replied:
I was sort of a assuming "private" functions/methods did not necessarily need docstrings. Although sometimes I have added them anyway, and I can here. But did I just invent that guideline, does everything need docstrings equally whether public or "private"?

pwinston replied:
I added a Parameters section and type annotations.

jni commented on napari/utils/perf/_patcher.py:
```diff
@@ -0,0 +1,155 @@
+        napari.utils.chunk_loader (success)
+        napari.utils.chunk_loader.ChunkLoader (failure)
+
+    The last one fails because ChunkLoader is a class not a module.
```
Needs a Parameters section

pwinston replied:
Added parameters section

jni commented on napari/utils/perf/_patcher.py:
```diff
@@ -0,0 +1,155 @@
+    -------
+    Tuple[types.ModuleType, str]
+        Where the module is the inner most imported module, and the string
+    is the rest of target_str that was not modules.
```
```suggestion
        is the rest of target_str that was not modules.
```

jni commented on napari/utils/perf/_patcher.py:
```diff
@@ -0,0 +1,155 @@
+        @wrapt.patch_function_wrapper(parent, callable)
+        def my_announcer(wrapped, instance, args, kwargs):
+            print(f"Announce {label}")
+            return wrapped(*args, **kwargs)
```
This won't render properly as is. For reference, here's a cheat sheet for ReStructuredText syntax, which is what's being used in these docstrings: https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#inserting-code-and-literal-blocks

```suggestion
    An example patch_func is::

        import wrapt
        def _my_patcher(parent: CallableParent, callable: str, label: str):
            @wrapt.patch_function_wrapper(parent, callable)
            def my_announcer(wrapped, instance, args, kwargs):
                print(f"Announce {label}")
                return wrapped(*args, **kwargs)
```

jni reviewed (commented):
Thanks @pwinston, much simpler now from my perspective! I've made a couple of extra notes around docstrings.

I'm unclear from your comment whether we should wait to merge for you to use my toolz suggestion to simplify the timer, or whether you'd rather move on to other things! Either is fine. (But toolz is awesome ;)

pwinston committed:
Change arg name and add type

pwinston committed:
Add attributes doc strring and related

Change pid/tid to process_id/thread_id
Add comment about "X" and "I"

pwinston committed:
Rename event.type to event.phase

This makes it clearer that it's a chrome://tracing thing and not just
a generic "type" thing.

pwinston committed:
Update napari/utils/perf/_patcher.py

Co-authored-by: Juan Nunez-Iglesias <juan.nunez-iglesias@monash.edu>

pwinston committed:
Update napari/utils/perf/_patcher.py

Co-authored-by: Juan Nunez-Iglesias <juan.nunez-iglesias@monash.edu>

pwinston committed:
Add Parameters

pwinston committed:
Added Parameters docstring

pwinston committed:
Remove "decorator" category string

I have not really noticed where this shows up in the UI so until we feel
it's useful best to leave it out.

pwinston committed:
Remove perf_func and perf_func_named

These are not needed now that we have the patcher.

pwinston committed:
Update comments

pwinston committed:
Merge branch 'master' into feature/perfmon2

pwinston committed:
Move perf_timer into _timers.py

Remove _utils.py since there was only 1 thing left in it.

pwinston committed:
Remove NAPARI_TRACE_FILE env var

Add new start_file_on_start config file option instead.

Using a separate env var was legacy, everything should be
controlled via the config file now.
